### PR TITLE
Inline Hints - ArgumentOutOfRangeException

### PR DESCRIPTION
**Pull request by akhera99**
Race condition caused by clearing out the cache in the TagsChanged event. The TagsChanged event can be run on a background thread, which may run while GetTags is being called. 
Switched to the BatchTagsChangedEvent since that is only raised on the same thread that created the TagAggregator.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1544392

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs [hunk 1]
```diff
@@ -71,7 +73,22 @@ internal sealed class InlineHintsTagger : ITagger<IntraTextAdornmentTag>, IDispo
             _formatMap = taggerProvider.ClassificationFormatMapService.GetClassificationFormatMap(textView);
             _hintClassification = taggerProvider.ClassificationTypeRegistryService.GetClassificationType(InlineHintsTag.TagId);
             _formatMap.ClassificationFormatMappingChanged += this.OnClassificationFormatMappingChanged;
-            _tagAggregator.TagsChanged += OnTagAggregatorTagsChanged;
+            _tagAggregator.BatchedTagsChanged += TagAggregator_BatchedTagsChanged;
+        }
+
+        private void TagAggregator_BatchedTagsChanged(object sender, BatchedTagsChangedEventArgs e)
+        {
+            _cacheSnapshot = null;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:81]: can you put a threading assert here?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:81]: and just generally doc what's going on.

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs [hunk 2]
```diff
@@ -71,7 +73,22 @@ internal sealed class InlineHintsTagger : ITagger<IntraTextAdornmentTag>, IDispo
...
+            _cache.Clear();
+            var mappingSpans = e.Spans;
+            foreach (var item in mappingSpans)
+            {
+                var spans = item.GetSpans(_buffer);
+                foreach (var span in spans)
+                {
+                    TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(span));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:89]: get and assign TagsChanged to a local first, check it for null, and do not proceed if null.

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs [hunk 3]
```diff
@@ -181,7 +189,7 @@ public IEnumerable<ITagSpan<IntraTextAdornmentTag>> GetTags(NormalizedSnapshotSp
 
         public void Dispose()
         {
-            _tagAggregator.TagsChanged -= OnTagAggregatorTagsChanged;
+            _tagAggregator.BatchedTagsChanged -= TagAggregator_BatchedTagsChanged;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:192]: doc.

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs [hunk 4]
```diff
@@ -5,11 +5,13 @@
 using System;
 using System.Collections.Generic;
 using System.Composition;
+using System.Linq;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:8]: I don't think these new usings are actually needed at this point?

**Review by jasonmalinowski — APPROVED** [review]

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs (hunk 4 above)
Comment (akhera99) [reply src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:8]: Interestingly they weren't greyed out, must be a bug.

**Comment by arkalyanms** [comment]
Thanks for the fix, approved for Preview 5!

**Review by arkalyanms — APPROVED** [review]

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs [hunk 5]
```diff
@@ -120,6 +121,12 @@ private TextFormattingRunProperties Format
             }
         }
 
+        private void InvalidateCache()
+        {
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:125]: would recommend threading assert here as well.

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs [hunk 6]
```diff
@@ -76,17 +68,27 @@ internal sealed class InlineHintsTagger : ITagger<IntraTextAdornmentTag>, IDispo
...
+            InvalidateCache();
+
             var mappingSpans = e.Spans;
             foreach (var item in mappingSpans)
             {
                 var spans = item.GetSpans(_buffer);
                 foreach (var span in spans)
                 {
-                    TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(span));
+                    var tagsChanged = TagsChanged;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs:87]: pull this out above *all* the foreachs.  if it is null you can insta-bail.

**Pull request merged**